net: l2: wifi: setting CMD only when net_if is up

MaochenWang1 opened the pull request:
Setting Wi-Fi cmd only when the net interface is up, avoid the case that driver deinit and net interface is down, then the L2 APIs interact with supplicant and driver, which may lead to CPU exception.

MaochenWang1 committed:
net: l2: wifi: setting CMD only when net_if is up

Setting Wi-Fi cmd only when the net interface is up, avoid the case
that driver deinit and net interface is down, then the L2 APIs interact
with supplicant and driver, which may lead to CPU exception.

Signed-off-by: Maochen Wang <maochen.wang@nxp.com>

MaochenWang1 commented:
Hi @jukkar @krish2718  please review

jukkar commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -733,6 +774,10 @@ static int wifi_iface_status(uint32_t mgmt_request, struct net_if *iface,
 		return -ENOTSUP;
 	}
 
+	if (!net_if_is_admin_up(iface)) {
+		return -ENETDOWN;
+	}
```
I thinks we should allow status command to get the current disconnected status when the interface is down.

MaochenWang1 replied:
done

jukkar commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -761,6 +806,10 @@ static int wifi_iface_stats(uint32_t mgmt_request, struct net_if *iface,
 		return -ENOTSUP;
 	}
 
+	if (!net_if_is_admin_up(iface)) {
+		return -ENETDOWN;
+	}
```
Also statistics could be shown for down interfaces like how `net stats` command does it.

jukkar commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -779,6 +828,10 @@ static int wifi_iface_stats_reset(uint32_t mgmt_request, struct net_if *iface,
 		return -ENOTSUP;
 	}
 
+	if (!net_if_is_admin_up(iface)) {
+		return -ENETDOWN;
+	}
```
Same for statistics reset

jukkar commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -1091,6 +1184,10 @@ static int wifi_get_version(uint32_t mgmt_request, struct net_if *iface,
 		return -ENOTSUP;
 	}
 
+	if (!net_if_is_admin_up(iface)) {
+		return -ENETDOWN;
+	}
```
I think we could allow the version command to print information even if interface is down.

jukkar reviewed (commented):
Some commands we could do even if interface is down.

MaochenWang1 commented:
> Some commands we could do even if interface is down.

Thanks, please review again @jukkar

krish2718 reviewed (dismissed):
While we are at it, we should fix long standing bug of checking for `!iface`, in case Wi-Fi driver init fails, shell crashes, I always ignored it :)

MaochenWang1 committed:
net: l2: wifi: add the check if dev is NULL

Add the check if dev is NULL (when iface is NULL, dev is NULL) when
setting Wi-Fi CMD, in case Wi-Fi driver init fails and shell crashes.

Signed-off-by: Maochen Wang <maochen.wang@nxp.com>

MaochenWang1 commented:
> While we are at it, we should fix long standing bug of checking for `!iface`, in case Wi-Fi driver init fails, shell crashes, I always ignored it :)
Hi @krish2718  update as your suggestion, please review

jukkar commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -545,6 +550,10 @@ static int wifi_neighbor_rep_complete(uint32_t mgmt_request, struct net_if *ifac
 	const struct wifi_mgmt_ops *const wifi_mgmt_api = get_wifi_api(iface);
 	struct wifi_scan_params params = {0};
 
+	if (wifi_mgmt_api == NULL || wifi_mgmt_api->candidate_scan == NULL) {
+		return -ENOTSUP;
+	}
+
```
How this change is related to the commit description which only talks about dev check?

MaochenWang1 replied:
Remove this change

MaochenWang1 commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -334,8 +334,13 @@ const char *wifi_ps_exit_strategy_txt(enum wifi_ps_exit_strategy ps_exit_strateg
 	const struct device *dev = net_if_get_device(iface);
-	struct net_wifi_mgmt_offload *off_api =
-			(struct net_wifi_mgmt_offload *) dev->api;
+	struct net_wifi_mgmt_offload *off_api = NULL;
```
Yes, update
